# Datasets II - Semantic Search
We will apply embeddings to this dataset to identify semantically similar issues using FAISS (Facebook AI Similarity Search)


In [2]:
from datasets import load_dataset

issues_dataset = load_dataset("lewtun/github-issues", split="train")
issues_dataset

C:\Users\Trevor_Kinsey\miniconda3\envs\hugging\lib\site-packages\huggingface_hub\repocard.py:105: UserWarning: Repo card metadata block was not found. Setting CardData to empty.
  warnings.warn("Repo card metadata block was not found. Setting CardData to empty.")


Generating train split: 0 examples [00:00, ? examples/s]

Dataset({
    features: ['url', 'repository_url', 'labels_url', 'comments_url', 'events_url', 'html_url', 'id', 'node_id', 'number', 'title', 'user', 'labels', 'state', 'locked', 'assignee', 'assignees', 'milestone', 'comments', 'created_at', 'updated_at', 'closed_at', 'author_association', 'active_lock_reason', 'pull_request', 'body', 'timeline_url', 'performed_via_github_app', 'is_pull_request'],
    num_rows: 3019
})

In [3]:
# filter out pull requests (they are labelled as issues)
issues_dataset = issues_dataset.filter(
    lambda x: (x["is_pull_request"] == False and len(x["comments"]) > 0)
)
issues_dataset

Filter:   0%|          | 0/3019 [00:00<?, ? examples/s]

Dataset({
    features: ['url', 'repository_url', 'labels_url', 'comments_url', 'events_url', 'html_url', 'id', 'node_id', 'number', 'title', 'user', 'labels', 'state', 'locked', 'assignee', 'assignees', 'milestone', 'comments', 'created_at', 'updated_at', 'closed_at', 'author_association', 'active_lock_reason', 'pull_request', 'body', 'timeline_url', 'performed_via_github_app', 'is_pull_request'],
    num_rows: 808
})

In [4]:
# keep only the columns we want
columns = issues_dataset.column_names
columns_to_keep = ["title", "body", "html_url", "comments"]
columns_to_remove = set(columns_to_keep).symmetric_difference(columns)
issues_dataset = issues_dataset.remove_columns(columns_to_remove)
issues_dataset


Dataset({
    features: ['html_url', 'title', 'comments', 'body'],
    num_rows: 808
})

To create our embeddings we’ll augment each comment with the issue’s title and body, since these fields often include useful contextual information. Because our comments column is currently a list of comments for each issue, we need to “explode” the column so that each row consists of an (html_url, title, body, comment) tuple. 

In Pandas we can do this with the DataFrame.explode() function, which creates a new row for each element in a list-like column, while replicating all the other column values. To see this in action, let’s first switch to the Pandas DataFrame format:

In [7]:
issues_dataset.set_format("pandas")
df = issues_dataset[:]
df.head()

html_url  \
0  https://github.com/huggingface/datasets/issues...   
1  https://github.com/huggingface/datasets/issues...   
2  https://github.com/huggingface/datasets/issues...   
3  https://github.com/huggingface/datasets/issues...   
4  https://github.com/huggingface/datasets/issues...   

                                               title  \
0                              Protect master branch   
1  Backwards compatibility broken for cached data...   
2  OSCAR unshuffled_original_ko: NonMatchingSplit...   
3  load_dataset using default cache on Windows ca...   
4  to_tf_dataset keeps a reference to the open da...   

                                            comments  \
0  [Cool, I think we can do both :), @lhoestq now...   
1  [Hi ! I guess the caching mechanism should hav...   
2  [I tried `unshuffled_original_da` and it is al...   
3  [Hi @daqieq, thanks for reporting.\r\n\r\nUnfo...   
4  [I did some investigation and, as it seems, th...   

                                                body  
0  After accidental merge commit (91c55355b634d0d...  
1  ## Describe the bug\r\nAfter upgrading to data...  
2  ## Describe the bug\r\n\r\nCannot download OSC...  
3  ## Describe the bug\r\nStandard process to dow...  
4  To reproduce:\r\n```python\r\nimport datasets ...

In [9]:
# explode out the comments so each has its own row with all the other fields intact
comments_df = df.explode("comments", ignore_index=True)
comments_df.head()

html_url  \
0  https://github.com/huggingface/datasets/issues...   
1  https://github.com/huggingface/datasets/issues...   
2  https://github.com/huggingface/datasets/issues...   
3  https://github.com/huggingface/datasets/issues...   
4  https://github.com/huggingface/datasets/issues...   

                                               title  \
0                              Protect master branch   
1                              Protect master branch   
2  Backwards compatibility broken for cached data...   
3  Backwards compatibility broken for cached data...   
4  Backwards compatibility broken for cached data...   

                                            comments  \
0                    Cool, I think we can do both :)   
1  @lhoestq now the 2 are implemented.\r\n\r\nPle...   
2  Hi ! I guess the caching mechanism should have...   
3  If it's easy enough to implement, then yes ple...   
4  Well it can cause issue with anyone that updat...   

                                                body  
0  After accidental merge commit (91c55355b634d0d...  
1  After accidental merge commit (91c55355b634d0d...  
2  ## Describe the bug\r\nAfter upgrading to data...  
3  ## Describe the bug\r\nAfter upgrading to data...  
4  ## Describe the bug\r\nAfter upgrading to data...

In [13]:
# Now make these into a dataset
from datasets import Dataset

comments_dataset = Dataset.from_pandas(comments_df)

# add a "comment_length" feature and remove short comments that are likely not useful
comments_dataset = comments_dataset.map(lambda x: {"comment_length": len(x["comments"].split())})
comments_dataset = comments_dataset.filter(lambda x: x["comment_length"] > 15)
comments_dataset

Map:   0%|          | 0/2964 [00:00<?, ? examples/s]

Filter:   0%|          | 0/2964 [00:00<?, ? examples/s]

Dataset({
    features: ['html_url', 'title', 'comments', 'body', 'comment_length'],
    num_rows: 2175
})

In [14]:
# concatenate the issue title, description, and comments together in a new text column
def concatenate_text(examples):
    return {
        "text": examples["title"]
        + " \n "
        + examples["body"]
        + " \n "
        + examples["comments"]
    }

comments_dataset = comments_dataset.map(concatenate_text)
comments_dataset

Map:   0%|          | 0/2175 [00:00<?, ? examples/s]

Dataset({
    features: ['html_url', 'title', 'comments', 'body', 'comment_length', 'text'],
    num_rows: 2175
})

# Create the embeddings 
To do this we must pick a suitable checkpoint. We can use the `sentence-transformers` model. 

This is an example of asymmetric semantic search because we have a short query whose answer we’d like to find in a longer document, like a an issue comment. Use the `multi-qa-mpnet-base-dot-v1` checkpoint

In [6]:
from transformers import AutoTokenizer, TFAutoModel

model_ckpt = "sentence-transformers/multi-qa-mpnet-base-dot-v1"
tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
model = TFAutoModel.from_pretrained(model_ckpt, from_pt=True)

Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFMPNetModel: ['embeddings.position_ids']
- This IS expected if you are initializing TFMPNetModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFMPNetModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).
All the weights of TFMPNetModel were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFMPNetModel for predictions without further training.


### cls pooling
Represent the semantic meaning of the sentence from the `[CLS]` token (the sequence's last hidden state)


In [7]:
def cls_pooling(model_output):
    return model_output.last_hidden_state[:, 0]

def get_embeddings(text_list):
    encoded_input = tokenizer(
        text_list, padding=True, truncation=True, return_tensors="tf"
    )
    encoded_input = {k: v for k, v in encoded_input.items()}
    model_output = model(**encoded_input)
    return cls_pooling(model_output)

In [20]:
# Get embedding for first example
embedding = get_embeddings(comments_dataset["text"][0])
embedding.shape

TensorShape([1, 768])

In [24]:
# now make a dataset of these embeddings
embeddings_dataset = comments_dataset.map(
    lambda x: {"embeddings": get_embeddings(x["text"]).numpy()[0]}
)

C:\Users\Trevor_Kinsey\miniconda3\envs\hugging\lib\site-packages\transformers\generation\tf_utils.py:465: UserWarning: `seed_generator` is deprecated and will be removed in a future version.
  warnings.warn("`seed_generator` is deprecated and will be removed in a future version.", UserWarning)


Map:   0%|          | 0/2175 [00:00<?, ? examples/s]

In [28]:
from datasets import load_from_disk
model_name = "github-issues"
embeddings_dataset.save_to_disk(model_name)
embeddings_dataset_reload = load_from_disk(model_name)
embeddings_dataset_reload

Saving the dataset (0/1 shards):   0%|          | 0/2175 [00:00<?, ? examples/s]

Dataset({
    features: ['html_url', 'title', 'comments', 'body', 'comment_length', 'text', 'embeddings'],
    num_rows: 2175
})

### FAISS Index
The basic idea behind FAISS is to create a special data structure called an index that allows one to find which embeddings are similar to an input embedding.

In [1]:
from datasets import load_from_disk
model_name = "github-issues"
embeddings_dataset = load_from_disk(model_name)
embeddings_dataset

Dataset({
    features: ['html_url', 'title', 'comments', 'body', 'comment_length', 'text', 'embeddings'],
    num_rows: 2175
})

In [2]:
embeddings_dataset.add_faiss_index(column="embeddings")

  0%|          | 0/3 [00:00<?, ?it/s]

Dataset({
    features: ['html_url', 'title', 'comments', 'body', 'comment_length', 'text', 'embeddings'],
    num_rows: 2175
})

We can now perform queries on this index by doing a nearest neighbor lookup with the Dataset.get_nearest_examples() function. Let’s test this out by first embedding a question as follows:

In [8]:
question = "How can I load a dataset offline?"
question_embedding = get_embeddings([question]).numpy()
question_embedding.shape

(1, 768)

Just like with the documents, we now have a 768-dimensional vector representing the query, which we can compare against the whole corpus to find the most similar embeddings:

In [9]:
scores, samples = embeddings_dataset.get_nearest_examples("embeddings", question_embedding, k=5)

In [10]:
import pandas as pd

samples_df = pd.DataFrame.from_dict(samples)
samples_df["scores"] = scores
samples_df.sort_values("scores", ascending=False, inplace=True)
samples_df

html_url  \
4  https://github.com/huggingface/datasets/issues...   
3  https://github.com/huggingface/datasets/issues...   
2  https://github.com/huggingface/datasets/issues...   
1  https://github.com/huggingface/datasets/issues...   
0  https://github.com/huggingface/datasets/issues...   

                                       title  \
4  Discussion using datasets in offline mode   
3  Discussion using datasets in offline mode   
2  Discussion using datasets in offline mode   
1  Discussion using datasets in offline mode   
0  Discussion using datasets in offline mode   

                                            comments  \
4  Requiring online connection is a deal breaker ...   
3  The local dataset builders (csv, text , json a...   
2  I opened a PR that allows to reload modules th...   
1  > here is my way to load a dataset offline, bu...   
0  here is my way to load a dataset offline, but ...   

                                                body  comment_length  \
4  `datasets.load_dataset("csv", ...)` breaks if ...              57   
3  `datasets.load_dataset("csv", ...)` breaks if ...              38   
2  `datasets.load_dataset("csv", ...)` breaks if ...             179   
1  `datasets.load_dataset("csv", ...)` breaks if ...              76   
0  `datasets.load_dataset("csv", ...)` breaks if ...              47   

                                                text  \
4  Discussion using datasets in offline mode \n `...   
3  Discussion using datasets in offline mode \n `...   
2  Discussion using datasets in offline mode \n `...   
1  Discussion using datasets in offline mode \n `...   
0  Discussion using datasets in offline mode \n `...   

                                          embeddings     scores  
4  [-0.47318100929260254, 0.24578352272510529, -0...  25.505011  
3  [-0.44908562302589417, 0.209506556391716, -0.0...  24.555534  
2  [-0.47164788842201233, 0.29022708535194397, -0...  24.148987  
1  [-0.4992600977420807, 0.22699761390686035, -0....  22.893990  
0  [-0.4902574121952057, 0.22889606654644012, -0....  22.406647

In [11]:
for _, row in samples_df.iterrows():
    print(f"COMMENT: {row.comments}")
    print(f"SCORE: {row.scores}")
    print(f"TITLE: {row.title}")
    print(f"URL: {row.html_url}")
    print("=" * 50)
    print()

COMMENT: Requiring online connection is a deal breaker in some cases unfortunately so it'd be great if offline mode is added similar to how `transformers` loads models offline fine.

@mandubian's second bullet point suggests that there's a workaround allowing you to use your offline (custom?) dataset with `datasets`. Could you please elaborate on how that should look like?
SCORE: 25.5050106048584
TITLE: Discussion using datasets in offline mode
URL: https://github.com/huggingface/datasets/issues/824

COMMENT: The local dataset builders (csv, text , json and pandas) are now part of the `datasets` package since #1726 :)
You can now use them offline
```python
datasets = load_dataset('text', data_files=data_files)
```

We'll do a new release soon
SCORE: 24.55553436279297
TITLE: Discussion using datasets in offline mode
URL: https://github.com/huggingface/datasets/issues/824

COMMENT: I opened a PR that allows to reload modules that have already been loaded once even if there's no internet.